### PR TITLE
[Feature] Filling education data in a user profile

### PR DESCRIPTION
**Pull request by nikitathebest**
**Link to the ticket:**
https://trello.com/c/VjVPbGqM/13-feature-filling-education-data-in-a-user-profile
**What it does:**
Adding and editing information about education
**DB changes:**
Created table educations
**UI changes:**
Created views for EducationsController

### CODE REVIEW

**Comment by gogochikago** [comment]
Please resolve conflicts!

**Comment by nikitathebest** [comment]
Conflicts resolved

File: app/controllers/educations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+# Controller for education
+class EducationsController < ApplicationController
+  def edit
+    @user = current_user
```
Comment (alexshgov) [inline app/controllers/educations_controller.rb]: No need to define an additional variable `@user`, you can use current_user from your views explicitly

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -6,6 +6,7 @@ def create
     @user = User.new(user_params)
     if @user.save
       ProfileCreateService.new(user_id: @user.id).call!
+      EducationCreateService.new(user_id: @user.id).call!
```
Comment (alexshgov) [inline app/controllers/users_controller.rb]: What if EducactionCreateService failed?

Consider [using transactions](https://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html) to make your actions atomic

File: app/services/education_create_service.rb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+  def call!
+    save_education!
+  end
+
+  private
+
+  def save_education!
+    education = Education.new(user_id: user_id)
+    education.save(validate: false)
```
Comment (alexshgov) [inline app/services/education_create_service.rb]: Validate false? Didn't get the idea, please clarify

File: db/migrate/20190527221008_add_details_to_educations.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+class AddDetailsToEducations < ActiveRecord::Migration[5.2]
+  def change
```
Comment (alexshgov) [inline db/migrate/20190527221008_add_details_to_educations.rb]: You can add all of these parameters in one migration: ._.create_ecucations.rb_, no need to pollute our migrations files

File: db/schema.rb [hunk 5]
```diff
@@ -26,6 +39,23 @@
     t.index ["user_id"], name: "index_profiles_on_user_id"
   end
 
+  create_table "skills", force: :cascade do |t|
```
Comment (alexshgov) [inline db/schema.rb]: This PR is about educations, what skills are doing here?

File: spec/services/education_create_service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+# rubocop:disable all
```
Comment (alexshgov) [inline spec/services/education_create_service_spec.rb]: Disable some metrics length for spec files in rubocop config

File: app/services/education_create_service.rb (hunk 3 above)
Comment (nikitathebest) [reply app/services/education_create_service.rb]: All the validation from model will be skipped when we use validate: false, so we will be able to create fields with nil values, and then edit them in the profile page by filling them with appropriate values

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -41,10 +42,19 @@ def destroy
...
+                                 %i[highschool faculty department year_of_start
+                                    year_of_end english])
   end
 
   def user_profile_id
     @user.profile.id
   end
+
+  def create_education
+    ActiveRecord::Base.transaction do
```
Comment (labs-ruby) [inline app/controllers/users_controller.rb]: Transaction block should be moved to the service

File: app/controllers/users_controller.rb [hunk 8]
```diff
@@ -6,6 +6,7 @@ def create
     @user = User.new(user_params)
     if @user.save
       ProfileCreateService.new(user_id: @user.id).call!
```
Comment (labs-ruby) [inline app/controllers/users_controller.rb]: So you created 2 services: `ProfileCreateService` and `EducationCreateService`.

Why not to create `UserCreateService` with the following code inside?
```ruby
class UserCreateService
   # ...
   def call
     ActiveRecord::Base.transaction do
      create_user!
      create_education!
      create_profile!
    end
   end
end
```

Comment (labs-ruby) [reply app/controllers/users_controller.rb]: For now, usage of services is performed incorrectly

File: db/schema.rb [hunk 9]
```diff
@@ -15,6 +15,19 @@
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
+  create_table "educations", force: :cascade do |t|
+    t.string "highschool"
+    t.string "faculty"
+    t.string "department"
+    t.integer "year_of_start"
+    t.integer "year_of_end"
```
Comment (labs-ruby) [inline db/schema.rb]: There's a type called `date`, use it instead of integer

File: app/controllers/educations_controller.rb [hunk 10]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class EducationsController < ApplicationController
+  def edit
+    @education = current_user.education
```
Comment (gogochikago) [inline app/controllers/educations_controller.rb]: This is not the best idea. Imagine that an administrator comes to your profile and wants to change your education. But the form will render his education (admin).

File: app/controllers/profiles_controller.rb [hunk 11]
```diff
@@ -4,6 +4,7 @@ class ProfilesController < ApplicationController
   def show
     @profile = Profile.find(params[:user_id])
     authorize @profile
+    @education = current_user.education
```
Comment (gogochikago) [inline app/controllers/profiles_controller.rb]: If I comes to another profile. What will I see?

File: app/views/educations/_show.html.slim [hunk 12]
```diff
@@ -0,0 +1,19 @@
+div id="education"
```
Comment (gogochikago) [inline app/views/educations/_show.html.slim]: For adaptability to different screen sizes, use bootstrap.

File: config/routes.rb [hunk 13]
```diff
@@ -17,12 +17,20 @@
   end
  
   resources :vacancies, param: :vac_id, except: %i[edit]
-  resources :users, param: :user_id, except: %i[index new]
+  # resources :users, param: :user_id, except: %i[index new]
```
Comment (gogochikago) [inline config/routes.rb]: Delete.

File: config/routes.rb [hunk 14]
```diff
@@ -17,12 +17,20 @@
...
   namespace :user do
     get '/vacancies', to: 'vacancies#index'
     resource :skill_level, only: %i[edit]
   end
 
+  resources :vacancies, param: :vac_id, except: %i[edit]
+  
+  resources :users, param: :user_id, except: %i[index new] do
+    member do
+      resource :education, only: %i[edit update]
```
Comment (gogochikago) [inline config/routes.rb]: Please, add namespace. For model and controller too.

File: spec/controllers/educations_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,58 @@
+require 'rails_helper'
+
+RSpec.describe EducationsController, type: :controller do
```
Comment (gogochikago) [inline spec/controllers/educations_controller_spec.rb]: What will happen if all these actions will be done by an unregistered user?

**Pull request merged**